Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,157 @@
+listBits' = unsafeCoerce listBits
+
+flipBit' :: KnownNat k => SMU.MVector k s Bit -> Mod k -> ST s ()
+flipBit' = unsafeCoerce (unsafeFlipBit :: MU.MVector s Bit -> Int -> ST s ())
+
+index' :: KnownNat k => SBMatrix k -> Mod k -> SBVector k
+index' = unsafeCoerce SV.index
+
+-- | It takes a random seed and a square singular matrix and it returns an
+-- elemnent of its kernel. It does not check if the matrix is singular.
+linearSolve :: KnownNat k => StdGen -> SBMatrix k -> DBVector k
+linearSolve seed matrix = linearSolveHelper 1 matrix randomVectors 1
+  where
+    -- The floating point number is the density of the random vectors.
+    randomVectors = getRandomDBVectors 0.1 seed
+
+linearSolveHelper :: KnownNat k => F2Poly -> SBMatrix k -> [DBVector k] -> Int -> DBVector k
+linearSolveHelper _ _ [] _ = error "Not enough random vectors"
+linearSolveHelper _ _ [_] _ = error "Not enough random vectors"
+linearSolveHelper previousPoly matrix (z : x : otherVecs) counter
+  -- If the algorithm does not find a solution, try another random vector @x@.
+  | potentialSolution == mempty && counter <= 5 = linearSolveHelper potentialMinPoly matrix (z : otherVecs) (counter + 1)
+  -- If the algorithm does not find a solution after five iterations,
+  -- most likely (>90%), it means that it picked a bad random vector @z@
+  -- in the image of the matrix. This changes @z@.
+  | potentialSolution == mempty && counter > 5  = linearSolveHelper 1 matrix otherVecs 1
+  -- This is a good solution.
+  | otherwise                                   = potentialSolution
+  where
+    potentialSolution = findSolution singularities matrix almostZeroVector
+    almostZeroVector = evaluate matrix z reducedMinPoly
+    (singularities, reducedMinPoly) = L.break unBit (U.toList $ unF2Poly potentialMinPoly)
+    -- Information gathered from the previous random vector @x@ is used in
+    -- the subsequent iteration.
+    potentialMinPoly = lcm previousPoly candidateMinPoly
+    candidateMinPoly = findCandidatePoly matrix z x
+
+-- Infinite lists of random DBVectors.
+getRandomDBVectors :: forall k. KnownNat k => Double -> StdGen -> [DBVector k]
+getRandomDBVectors density gen = go $ randomRs (0, 1) gen
+  where
+    numberOfColumns = naturalToInt $ natVal (Proxy :: Proxy k)
+    go :: KnownNat k => [Double] -> [DBVector k]
+    go list = newVector `seq` newVector : go backOfList
+      where
+        newVector = DBVector $ fromJust $ SU.fromList $ map (\d -> Bit (d < density)) frontOfList
+        (frontOfList, backOfList) = L.splitAt numberOfColumns list
+
+-- The input is a matrix @B@ and random vectors @z@ and @x@. It returns the
+-- sequence @x A ^ k z@ as k runs from 0 to the size of the matrix.
+generateData :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+generateData matrix z x = toF2Poly $ U.fromList $ reverse $ map (x `dot`) matrixPowers
+  where
+    matrixPowers = take (((*2) . naturalToInt . natVal) matrix) $ L.iterate (matrix `mult`) z
```

Comment:
It makes sense to factor out a helper `intVal = naturalToInt . natVal` and use it in three places.